doc/dev/adr/0020-error-modeling-headers.md

```diff
@@ -0,0 +1,103 @@
+
+## Code Example
+
+### Command executor side
+
+```csharp
+public ExtendedResponse<IncrementResponsePayload> Increment(IncrementRequestPayload request)
+{
+    if (request.IncrementValue < 0)
+    {
+        var response = new ExtendedResponse<IncrementResponsePayload>()
+        {
+            Response = new IncrementResponsePayload { CounterResponse = _counter },
+        };
+
+        // Specify error code, but no error payload  
+        response.WithApplicationError("negativeValue");
+
+        // Or you can specify error code and error payload
+        response.WithApplicationError(
+            "negativeValue",
+            new CustomErrorPayload() { ArgumentValue = request.IncrementValue },
+            new ErrorPayloadJsonSerializer(new TestEnvoys.Utf8JsonSerializer()));
+
+        return response;
+    }
+
+    // happy path omitted
+}
+```
+
+### Command invoker side
+
+```csharp
+public void main()
+{
+    await using MqttSessionClient mqttInvoker = await ClientFactory.CreateSessionClientFromEnvAsync();
+    await using CounterClient counterClient = new CounterClient(applicationContext, mqttInvoker);
+
+    IncrementRequestPayload payload = new IncrementRequestPayload
+    {
+        IncrementValue = -1
+    };
+
+    var response = await counterClient.Increment(executorId, payload).WithMetadata();
+    
+    // Check the RPC response for an application error
+    if (response.TryGetApplicationError(SomeJsonSerializer, out string? errorCode, out CustomErrorPayload? errorPayload))
+    {
+        // use the error code and strongly typed custom error payload type as wanted.
+    }
+
+    // Alternatively just fetch the error code if you don't want to deserialize the payload yet.
+    if (response.TryGetApplicationError(out string? errorCode))
+    {
+        // ...
+    }
+}
+```
+
+### Enforcement
+
+Note that, while this decision creates a standardized way of communicating application errors, our SDK will **not** enforce that users communicate application errors with this standard. Users will still be able to model application errors in response payloads or in custom user properties if they prefer. There is no way to force users to communicate errors with this standard pattern.
+
+## Samples
+
+We want to establish this pattern as the standard for communicating RPC application errors. To do so, we must either update or add a sample per-language that demonstrates handling at least one application error.
+
+For instance, we would make the counter service return an application error if the invocation specifies a negative value to increment by. This change won't include any DTDL level changes yet.
+
+## Open Questions
+
+- MQTT user properties are strings, not bytes, so is there an issue with using UTF8 encoding to convert a serialized payload to a string? Does this work for the serializers we already support (protobuf, avro, raw, etc.)
```

Comment:
The formats of AVRO, Protobuf, and raw are binary, so these would need to be Base64-encoded if we want to pack them into MQTT header values.